arex-schedule-web-api/src/main/java/com/arextest/schedule/web/controller/ReplayLocalController.java

```diff
@@ -0,0 +1,92 @@
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.Resource;
+import javax.validation.Valid;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.collections4.CollectionUtils;
+import org.springframework.http.MediaType;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+/**
+ * @author wildeslam.
+ * @create 2023/11/15 17:23
+ */
+@Slf4j
+@Controller
+@RequestMapping(path = "/api/replay/local/", produces = {MediaType.APPLICATION_JSON_VALUE})
+public class ReplayLocalController {
+
+  private static final String SUCCESS_DESC = "success";
+
+  @Resource
+  private LocalReplayService localReplayService;
+  @Resource
+  private PlanProduceService planProduceService;
+  @Resource
+  private ReplayPlanController replayPlanController;
+
+  @PostMapping(value = "/queryCaseId")
+  @ResponseBody
+  public CommonResponse queryCaseId(@Valid @RequestBody BuildReplayPlanRequest request) {
+    try {
+      replayPlanController.fillOptionalValueIfRequestMissed(request);
+      return localReplayService.queryReplayCaseId(request);
+    } catch (Exception e) {
+      LOGGER.error("queryCaseId error: {} , request: {}", e.getMessage(), request, e);
+      return CommonResponse.badResponse("queryCaseId error！" + e.getMessage(),
+          new BuildReplayPlanResponse(BuildReplayFailReasonEnum.UNKNOWN));
+    } finally {
+      planProduceService.removeCreating(request.getAppId(), request.getTargetEnv());
+    }
+  }
+
+  @PostMapping(value = "/queryReplaySenderParameters")
+  @ResponseBody
+  public CommonResponse queryCases(@Valid @RequestBody QueryReplaySenderParametersRequest request) {
+    if (CollectionUtils.isEmpty(request.getCaseIds())) {
+      return CommonResponse.badResponse("No caseId!");
+    }
+    QueryReplaySenderParametersResponse response = localReplayService.queryReplaySenderParameters(request);
+    String desc = "";
+    if (Objects.equals(response.getReplaySenderParametersMap().size(), request.getCaseIds().size())) {
+      desc = "QueryCases success!";
+    } else {
+      desc = "Some cases are missing.";
+    }
+    return CommonResponse.successResponse(desc, response);
+  }
+
+  @PostMapping(value = "/preSend")
+  @ResponseBody
+  public CommonResponse preSend(@Valid @RequestBody preSendRequest request) {
+    boolean success = localReplayService.preSend(request);
+    if (success) {
+      return CommonResponse.successResponse(SUCCESS_DESC, success);
+    } else {
+      return CommonResponse.badResponse("SendLimiter break!", success);
+    }
+  }
+
+  @PostMapping(value = "/postSend")
+  @ResponseBody
+  public CommonResponse postSend(@Valid @RequestBody PostSendRequest request) {
+    CompletableFuture.runAsync(() -> localReplayService.postSend(request));
```

Comment:
Use custom threadpool
It is recommended that asynchronous logic be encapsulated in the service layer

Reply:
done